packages/ses/src-xs/index.js

```diff
@@ -0,0 +1,463 @@
+          moduleMapHook(specifier) ??
+          (await importHook(specifier));
+        mapDelete(descriptors, specifier);
+        descriptor = adaptModuleDescriptor(descriptor, specifier, compartment);
+        return descriptor;
+      };
+      nativeOptions = {
+        ...nativeOptions,
+        resolveHook,
+        importHook: nativeImportHook,
+        loadHook: nativeImportHook,
+      };
+    }
+
+    if (importNowHook) {
+      /** @param {string} specifier */
+      const nativeImportNowHook = specifier => {
+        let descriptor =
+          mapGet(descriptors, specifier) ??
+          moduleMapHook(specifier) ??
+          importNowHook(specifier);
+        mapDelete(descriptors, specifier);
+        descriptor = adaptModuleDescriptor(descriptor, specifier, compartment);
+        return descriptor;
+      };
+      nativeOptions = {
+        ...nativeOptions,
+        resolveHook,
+        importNowHook: nativeImportNowHook,
+        loadNowHook: nativeImportNowHook,
+      };
+    }
+
+    const compartment = new NativeCompartment(nativeOptions);
+
+    const nativeGlobalThis = nativeGetGlobalThis(compartment);
+
+    const nativeEval = nativeGlobalThis.eval;
+
+    weakmapSet(privateFields, compartment, {
+      name,
+      transforms,
+      delegateNative,
+      noNamespaceBox,
+      nativeEval,
+      descriptors,
+    });
+
+    const shimOptions = {
+      ...options,
+      __noNamespaceBox__: true,
+      __options__: true,
+    };
+
+    uncurryThis(ShimCompartment)(compartment, shimOptions);
+
+    const shimGlobalThis = shimGetGlobalThis(compartment);
+
+    const ChildCompartment = nativeEvaluate(compartment, compartmentShim)(
+      compartmentShim,
+      commons,
+      nativeGlobalThis.Compartment,
+      shimGlobalThis.Compartment,
+      maybeHarden,
+    );
+
+    defineProperty(nativeGlobalThis, 'Compartment', {
+      value: ChildCompartment,
+      writable: true,
+      configurable: true,
+      enumerable: false,
+    });
+
+    defineProperty(shimGlobalThis, 'Compartment', {
+      value: ChildCompartment,
+      writable: true,
+      configurable: true,
+      enumerable: false,
+    });
+
+    maybeHarden(compartment);
+    return compartment;
+  }
+
+  return maybeHarden(Compartment);
+};`;
+
+// Adapt the start compartment's native Compartment to the SES-compatibility
+// adapter.
+// Before Lockdown, the Compartment constructor in transitive child
+// Compartments is not (and cannot be) hardened.
+const noHarden = object => object;
+globalThis.Compartment = FERAL_EVAL(compartmentShim)(
+  compartmentShim,
+  commons,
+  NativeCompartment,
+  ShimCompartment,
+  noHarden,
+);
+
+globalThis.lockdown = options => {
+  const hardenIntrinsics = repairIntrinsics(options);
+  hardenIntrinsics();
+  // Replace global Compartment with a version that is hardened and hardens
+  // transitive child Compartment.
+  globalThis.Compartment = FERAL_EVAL(compartmentShim)(
```

Comment:
This both reevaluates `compartmentShim` and reapplies its completion value, which is the big function defined by inline text above, to similar arguments, with the main (only?) difference being `noHarden` vs `harden`. While this is all fine, so no change necessarily suggested, it does make me wonder whether we could evaluate the function only once and have all needed differences come only from the application arguments?

Reply:
I’ve posted a change that eliminates the eval entirely. It’s now just a function `adaptCompartmentConstructors(NativeCompartmentConstructor, ShimCompartmentConstructor, maybeHarden)`. I look forward to possibly remembering why the eval was necessary, but it probably wasn’t actually.